### DISCUSSION TITLE
More tutorials would be great!

### DISCUSSION THREAD

**Discussion by Robsteranium** [thi-ng/umbrella#389] [Ideas]
For the most part, umbrella is documented with generated API docs and examples with some of the main packages enjoying longer readmes.

API docs are a useful reference if you already know how to use the package but don't help to introduce the tools.

Examples help to show how things fit together but some of them are really quite advanced and dense. When they combine multiple packages it's hard to start learning any given package as you have to first learn another one.

Those longer readmes that do exist are great - particular the pared-down examples - but they too can be quite dense and generally assume that the reader already knows the concepts involved.

By contrast the "How to UI in 2018" blog posts are really helpful because they don't just introduce umbrella packages but they also explain the background to the techniques with motivating examples and build up the solutions gradually.

Clearly these posts are _a lot_ of work so I'm not surprised theres not more (e.g. a "How to UI in 2023" might introduce rdom). Indeed I think the focus on new features and non-trivial examples is right - particular for @postspectacular. I don't think it's necessarily the responsibility of the umbrella project to explain the underlying concepts (and certainly not _all_ of them). That said I think some more tutorials could leverage all this hard work by making it accessible to a wider range of developers (etc).

This feels like an area where other contributors could get involved. The prospect is quite daunting however!

It'd be interesting to know if others felt this would be worthwhile and, if so, had any specific ideas for topics...

**Comment by nkint** [comment]
Documentation is an important part and it's an essential step. Imho it represents a building block, nothing else can be done without it and it is the result of a big effort.

Said that, I also think that an advance concept like transducers become natural after [of-umbrellas-transducers-reactive-streams-mushrooms](https://medium.com/@thi.ng/of-umbrellas-transducers-reactive-streams-mushrooms-pt-1-a8717ce3a170).
This 3 part blog-post is well written and easy to follow: it denotes a good skill in writing/teaching other than tech ones 👏

Last but not least also t[he stream](https://www.youtube.com/watch?v=hEC_qbUXDo8) on youtube is a really good way to understand how to combine multiple packages.

> This feels like an area where other contributors could get involved.

👍

I think it would be great to have more tutorials!
@Robsteranium what topics would you be interested in?

Random thought: Can this be a good moment to revamp [Algoapril](https://github.com/algoapril/algoapril-2022) and pick a topic for  new (community-made) tutorial?

**Comment by Robsteranium** [comment]
Yeah that's exactly the series of blog posts I was referring to - afaik they're the only ones about umbrella. They're great because they introduce abstract techniques like cellular automata/ signed distance fields in the context of an application - the Shroomania game. They also take care to build things up in steps and don't assume much context (except perhaps ES6).

I've not found time to watch the whole of that stream yet. I'm interested in WebGL/ shaders but have been working through other guides to understand the foundational concepts first. I had originally tried searching thi.ng for "3D" but it was only after I'd first learned about homogenous coordinates (etc) that I tracked down `@thi.ng/matrices/perspective`. There's *a lot* of tools available here if you know what to look for... but if you don't know what the tool is already you mightn't ever find it.

I've been using hdom for most things but I'm keen to get my head around rdom/ rstream since that's where the development effort is focussed - indeed it looks like this makes the data binding/ flow more explicit.

Judging by the introductory discussion thread (#374) the interest here is overwhelmingly in generative art so that's an obvious topic to focus on. 

Perhaps it would make sense to create tutorials for some of the existing examples not least as the hard work has already been done!
